java/feathub-connectors/flink-connectors/flink-connector-redis/src/main/java/com/alibaba/feathub/flink/connectors/redis/sink/RedisSinkFunction.java

```diff
@@ -48,86 +56,77 @@
         super.open(parameters);
 
-        this.indexBuffer = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN);
-        this.timestampBuffer = ByteBuffer.allocate(8).order(ByteOrder.BIG_ENDIAN);
-
         client = JedisClient.create(this.config);
-
-        if (timestampFieldIndex >= 0) {
-            evalScriptSHA = client.scriptLoad(this.evalScript);
-        }
     }
 
     @Override
     public void invoke(RowData data, Context context) {
-        byte[] originalKey = data.getBinary(keyFieldIndex);
-        byte[] key = new byte[keyPrefix.length + originalKey.length];
-        System.arraycopy(keyPrefix, 0, key, 0, keyPrefix.length);
-        System.arraycopy(originalKey, 0, key, keyPrefix.length, originalKey.length);
-
-        Map<byte[], byte[]> value = new HashMap<>();
-        int arity = data.getArity();
-        for (int i = 0; i < arity; i++) {
-            if (timestampFieldIndex == i) {
-                // timestamp values have specific serialization strategy
-                continue;
-            }
-
-            if (keyFieldIndex == i) {
-                // don't store key values
-                continue;
-            }
-
-            indexBuffer.putInt(0, i);
-            value.put(indexBuffer.array(), data.getBinary(i));
+        List<String> keyValues = new ArrayList<>();
+        for (int i : keyFieldIndices) {
+            keyValues.add(getString(data, i, fieldTypes[i]));
         }
-
-        if (timestampFieldIndex < 0) {
-            this.client.hset(key, value);
-        } else {
-            timestampBuffer.putLong(0, data.getLong(timestampFieldIndex));
-
-            List<byte[]> arguments = new ArrayList<>();
-            arguments.add(key);
-            arguments.add(timestampBuffer.array());
-            for (Map.Entry<byte[], byte[]> entry : value.entrySet()) {
-                arguments.add(entry.getKey());
-                arguments.add(entry.getValue());
+        String keyPrefix =
+                namespace + REDIS_KEY_DELIMITER + String.join(ROW_KEY_DELIMITER, keyValues);
+
+        for (int i : valueFieldIndices) {
+            String key = keyPrefix + REDIS_KEY_DELIMITER + fieldNames[i];
+            DataType dataType = fieldTypes[i];
+            if (dataType instanceof KeyValueDataType) {
+                client.del(key);
+                client.hmset(key, getMap(data.getMap(i), (KeyValueDataType) dataType));
+            } else if (dataType instanceof CollectionDataType) {
+                client.del(key);
+                client.rpush(
+                        key,
+                        getList(data.getArray(i), (CollectionDataType) dataType)
+                                .toArray(new String[0]));
+            } else {
+                client.set(key, getString(data, i, dataType));
             }
-
-            this.client.evalsha(this.evalScriptSHA, Collections.singletonList(key), arguments);
         }
+
+        client.flush();
```

Comment:
`SinkFunction#invoke` does not have the flush semantics. Would it be better to execute `sync()` only in `SinkFunction#finish`?

Reply:
Invoking `flush()` only at `SinkFunction#finish` can work only when the input stream is bounded. As it has been avoided to re-instantiate the pipeline for every record as described in the first comment, I think flushing out data at `SinkFunction#invoke` would not have that much performance overhead. What do you think?

Reply:
Calling flush() on every record typically will reduce throughput because there might be much less data sent in each round-trip time. After all, there is no need to provide an API to explicitly flush data if there is no latency/throughput tradeoff.

Flink is typically used for use-case with nearline latency with high throughput requirement. So it seems better not to call flush() for every record. We need to call flush() as part of close().